src/scalems/exceptions.py

```diff
@@ -0,0 +1,53 @@
+"""Exceptions thrown by SCALE-MS are catchable as scalems.ScaleMSException.
+
+Additional common exceptions are defined in this module.
+scalems submodules may define additional exceptions, but all will be derived
+from exceptions specified in scalems.exceptions.
+"""
+
```

Comment:
What is the advantage of defining custom exceptions?  Now a python knowledgeable user needs to learn (or lookup) new exception names for known things (`ValueError` vs. `InvalidArgumentError`).  Also, since the ScaleMS implementation is not wrapping/translating all external call, should the user now catch both the standard errors which might be raised from imported modules and ScaleMS specific errors from within this module?

I would see an argument of subtyping known exception types.  Like
```py
class DataShapeError(ValueError):
    ...
```
may provide the user with additional information, but still would allow to capture known and expected exception types.

Reply:
It seems like most "best practices" suggest a root exception in C++ or Python to clarify exceptions emitted by a particular library.

Some Python packages have started to inherit from both the package root exception _and_ similar built-in exceptions, and I could not find anyone advising against such a practice. I would be amenable to that.

However, in particular with these exceptions, the mapping to built-in exceptions isn't particularly clear. It is potentially difficult to decide whether something is a ValueError or a TypeError. I think it is more useful to provide information about _what_ is wrong with the provided object.

I don't think anyone is going to try to catch a `ValueError` unless the documentation for a function says it may throw a ValueError, and the user still needs to look up why it might through a ValueError.

> should the user now catch both the standard errors which might be raised from imported modules and ScaleMS specific errors from within this module?

Users generally either catch
* no exceptions
* specific exceptions, per documentation or past experience, or
* a root exception

Catching `BaseException` is almost useless and will even raise warnings in linters, but if someone has such a use case, we should at least provide them a chance to catch our base exception instead.

If an exception not derived from our root exception reaches the user, it should be really clear that they are using Python wrong, or it indicates a bug in our software.

If an exception derived from our root exception reaches the user, it should be very clear whether they are using our software wrong, or whether there is a bug in our software.

Reply:
> the ScaleMS implementation is not wrapping/translating all external call

For the most part, I believe it should be. With some potential caveats for clear Python usage errors, if a user is calling scalems as documented, they should only encounter the documented exceptions or (please) report a bug to us. Can you think of an exception? (no pun intended)

Actually, an important exception is when we need to fulfill a Python protocol, such as when code is _supposed_ to raise `AttributeError`, `StopIteration`, or `NotImplementedError`. Where we need to do this in a ScaleMS component, it is good to know that we have provided the user with an alternative to catching the built-in `Exception` so that no one has to worry about whether we are breaking the generator protocol, data descriptor protocol, or abstract base class machinery or something.

Reply:
Some more built-in exceptions that have special meaning are `IndexError`, `KeyError`, and `TypeError`, which should be emitted under specific circumstances by operator overloads and such. I'm open to the idea of something like `class LookupError(KeyError, ScaleMSException)`, but I am not sure if that adds utility.

If we code a `raise` in a `__getitem__()` implementation when a "key is not found in the set of existing keys", we should be raising a `KeyError`. If we raise a subclass of `KeyError`, though, it should also be a subclass of `ScaleMSException`, I believe.

Reply:
To further illustrate potential confusion, note that mismatched function signatures usually raise `TypeError`, not `ValueError`. I'm not sure how well-known that is. But it is hard to decide what you should raise from a function that accepts some variation of `*args, **kwargs` if the developer can't figure out what to do with the provided arguments. 

There isn't an easy answer, as far as I can tell, because it often depends on what the function is expected to do with the arguments. 

But by conventional reasoning, it seems to me that the appropriate built-in exception to raise in place of `InvalidArgumentError` could change as SCALE-MS evolves. That doesn't seem helpful, unless users could continue to catch a useful base exception. 

I'm not totally sold on `InvalidArgumentError`, but I was prompted to include it because the conventions for when to use `ValueError` and when to use `TypeError` did not seem helpful for saying "You are using Python right, but you're using SCALE-MS wrong."

Reply:
> There isn't an easy answer...

nope.  I am not sure about multiple inheritence either.  While it gives some middle ground between native and custom exceptions, I also doubt it adds clarity and utility.

> I'm not totally sold on InvalidArgumentError, but I was prompted to include it because the conventions for when to use ValueError and when to use TypeError did not seem helpful for saying "You are using Python right, but you're using SCALE-MS wrong."

Hmm, I am a bit surprised by that argument.  Usually it should be clear cut if a type is wrong (the passed parameter cannot be cast to the expected type) and when the value is wrong (the passed parameter has the correct type, but the value is unexpected and cannot be handled).  Why wouldn't the same hold for `args` and `kwargs`?  They also need to be expanded and interpreted eventually...

Reply:
`TypeError` indicates there is something wrong with an interface, while `ValueError` indicates something else wrong with the object. I don't think this distinction is meaningful to users who are taking the output of one command as the input to another, and I think it is a burden for both developers and users if we try to rigorously update these exceptions as lower level details change.

If the parameter is a container, but the argument contains the wrong payload, do we raise a `TypeError` when we try to use the object in the container, or a `ValueError` when we validate the function arguments?

If an argument is an object with a composed interface, and doesn't implement a needed hook, is that a `TypeError` or a `ValueError`? Or an `AttributeError`? Or a `NotImplementedError`? Does it matter to the user how we are implementing the composition?

My heart rate is starting to increase just thinking about it! :-\

I can take `InvalidArgumentError` out of the main exceptions module, but I think I will have to use more module-local `InvalidDataSource` or `InvalidWidgetFactory` exceptions, to avoid the precedence of a universal `InvalidArgumentError` exception, at least until the interfaces stabilize a bit more. I'll also make another pass to reduce overspecification, adding exceptions only as they have a demonstrated use so we can discuss the individually, and consider alternatives.